arangod/RocksDBEngine/RocksDBIterators.cpp

```diff
@@ -333,3 +333,133 @@ void RocksDBSortedAllIterator::reset() {
+  bool has_more = _iterator->Valid();
+  while (count > 0 && has_more) {
+    has_more = next([&count,&skipped](rocksdb::Slice const&, rocksdb::Slice const&){ --count; ++skipped; }, count /*gets copied*/);
+  }
+  return has_more;
+}
+
+bool RocksDBGenericIterator::seek(rocksdb::Slice const& key) {
+  if (_reverse) {
+    _iterator->SeekForPrev(key);
+  } else {
+    _iterator->Seek(key);
+  }
+  //TRI_ASSERT(_iterator->Valid()); - can be empty
+  return hasMore();
+}
+
+
+bool RocksDBGenericIterator::next(GenericCallback const& cb, size_t limit) {
+  // @params
+  // limit - maximum number of documents
+
+  TRI_ASSERT(limit > 0);  // Someone called with limit == 0. Api broken
+  if (limit == 0) {
+    // No limit no data, or we are actually done. The last call should have
+    // returned false
+    return false;
+  }
+
+  while (limit > 0 && hasMore()){
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+    TRI_ASSERT(_bounds.objectId() == RocksDBKey::objectId(_iterator->key()));
+#endif
+
+    cb(_iterator->key(),_iterator->value());
+    --limit;
+    if (_reverse) {
+      _iterator->Prev();
+    } else {
+      _iterator->Next();
+    }
+
+  }
+
+  return hasMore();
+}
+
+RocksDBGenericIterator arangodb::createPrimaryIndexIterator(transaction::Methods* trx
+                                                          ,LogicalCollection* col
+                                                          ){
+  TRI_ASSERT(col != nullptr);
+  TRI_ASSERT(trx != nullptr);
+
+  auto* mthds = RocksDBTransactionState::toMethods(trx);
+
+  rocksdb::ReadOptions options = mthds->readOptions(); // intentional copy of the read options
+  TRI_ASSERT(options.snapshot != nullptr); // trx must contain a valid snapshot
+  TRI_ASSERT(options.prefix_same_as_start);
+  options.fill_cache = false;
+  options.verify_checksums = false;
+
+  auto index = col->lookupIndex(0); //RocksDBCollection->primaryIndex() is private
+  TRI_ASSERT( index->type() == Index::IndexType::TRI_IDX_TYPE_PRIMARY_INDEX );
+  auto primaryIndex = static_cast<RocksDBPrimaryIndex*>(index.get());
+
+  auto bounds(RocksDBKeyBounds::PrimaryIndex(primaryIndex->objectId()));
+  auto iterator =  RocksDBGenericIterator(options, bounds);
+
+  TRI_ASSERT(iterator.bounds().objectId() == primaryIndex->objectId());
+  TRI_ASSERT(iterator.bounds().columnFamily() == RocksDBColumnFamily::primary());
+  return iterator;
+}
+
+RocksDBGenericIterator arangodb::createDocumentIterator(transaction::Methods* trx
+                                                          ,LogicalCollection* col
+                                                          ){
+  TRI_ASSERT(col != nullptr);
+  TRI_ASSERT(trx != nullptr);
+
+  auto* mthds = RocksDBTransactionState::toMethods(trx);
+
+  rocksdb::ReadOptions options = mthds->readOptions(); // intentional copy of the read options
+  TRI_ASSERT(options.snapshot != nullptr); // trx must contain a valid snapshot
+  TRI_ASSERT(options.prefix_same_as_start);
+  options.fill_cache = false;
```

Comment:
doesn't populate the cache anymore...
Previous implementation of the AllIterator did.

Reply:
fixed